### PR TITLE
Replace tile metadata files with layouts database

### PR DESCRIPTION
**Pull request by kirstenmg**
Instead of traversing entire resource file system to load tile layouts, store tile layouts in a single database.
LayoutDatabase handles updates to the database, and operations that dealt with the metadata files have been replaced with calls to the corresponding LayoutDatabase functions.

### CODE REVIEW

File: tasm/tiles/include/LayoutDatabase.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+class LayoutDbConfiguration {
+public:
+    static const std::experimental::filesystem::path &LayoutDbPath() {
+        return EnvironmentConfiguration::instance().defaultLayoutDatabasePath();
+    }
+};
+
+class LayoutDatabase {
+public:
+    static std::shared_ptr<LayoutDatabase> &instance() {
```
Comment (maureendaum) [inline tasm/tiles/include/LayoutDatabase.h]: Because this is a pointer, we don't need to return a reference. However, if we want to enforce that callers of this can't change it, we can add some `const` annotations.
Option 1: return a const reference `const LayoutDatabase&`
Option 2: return a shared_ptr to a constant object: `std::shared_ptr<const LayoutDatabase>`
Option 3: return a constant shared_ptr to a non-constant object: `const std::shared_ptr<LayoutDatabase>`
Option 4: return a constant shared_ptr to a constant object: `const std::shared_ptr<const LayoutDatabase>`
Option 4 means that a caller can't re-assign the pointer, and they also can't modify the pointed-to object. 

I haven't been good about putting `const` in or around pointers in the rest of the project, so I don't feel super strongly about having to add them here.

File: tasm/tiles/include/LayoutDatabase.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        if (instance_ == nullptr) {
```
Comment (maureendaum) [inline tasm/tiles/include/LayoutDatabase.h]: This can be shortened to `if (!instance_)`

File: tasm/tiles/include/LayoutDatabase.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            instance_ = std::make_shared<LayoutDatabase>(LayoutDatabase());
```
Comment (maureendaum) [inline tasm/tiles/include/LayoutDatabase.h]: I'm not positive, but I think that the compiler may be able to figure out how to construct the LayoutDatabase based on the template, so you may be able to shorten this to `std::make_shared<LayoutDatabase>();`

File: tasm/tiles/include/TiledVideoManager.h [hunk 4]
```diff
@@ -5,12 +5,14 @@
...
 #include "Video.h"
 #include <mutex>
+#include "LayoutDatabase.h"
 
 namespace tasm {
 class TiledVideoManager {
 public:
     TiledVideoManager(std::shared_ptr<TiledEntry> entry)
             : entry_(entry),
+              layouts_(LayoutDatabase::instance()),
```
Comment (maureendaum) [inline tasm/tiles/include/TiledVideoManager.h]: This is a nit and feel free to ignore, but `layouts_` makes me think that it is some sort of container (because of the plural-ness). What about `layoutDB_`?

File: tasm/tiles/src/LayoutDatabase.cc [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+    const char *createIndex = "CREATE INDEX heights_index ON heights(video, id)";
+    result = sqlite3_exec(db_, createIndex, NULL, NULL, &error);
+    if (result != SQLITE_OK) {
+        std::cerr << "Error creating heights index: " << sqlite3_errmsg(db_) << std::endl;
+        sqlite3_free(error);
+    }
+}
+
+std::vector<int> LayoutDatabase::tileLayoutIdsForFrame(const std::string &video, unsigned int frameNumber) const {
+    std::string selectLayoutIds = "SELECT id FROM layouts WHERE video = ? AND ? BETWEEN first_frame AND last_frame";
```
Comment (maureendaum) [inline tasm/tiles/src/LayoutDatabase.cc]: Because we are presumably going to be executing this query multiple times, it may make sense initialize it as a prepared statement, and then do bind/reset for each invocation. An example of this is in `SemanticIndexSQLite::initializeStatements(), ::destroyStatements(), ::addMetadata()`

(Ditto for the rest of the queries in this class)

File: tasm/tiles/src/LayoutDatabase.cc [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    unsigned int numberOfCols = 0;
+    if ((result = sqlite3_step(select)) == SQLITE_ROW) {
+        numberOfRows = sqlite3_column_int(select, 0);
+        numberOfCols = sqlite3_column_int(select, 1);
+    }
+
+    ASSERT_SQLITE_DONE(sqlite3_step(select));
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+
+    std::vector<unsigned int> heightsOfRows = heightsForId(video, id);
```
Comment (maureendaum) [inline tasm/tiles/src/LayoutDatabase.cc]: If you get tired of writing these explicit types for the return values of functions, you can use `auto heightsOfRows = …`.

File: tasm/tiles/src/LayoutDatabase.cc [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+
+    std::vector<unsigned int> widthsOfColumns = widthsForId(video, id);
+
+    return std::make_shared<TileLayout>(TileLayout(numberOfCols, numberOfRows, widthsOfColumns, heightsOfRows));
```
Comment (maureendaum) [inline tasm/tiles/src/LayoutDatabase.cc]: Because we don't use heightsOfRows and widthsOfColumns, you could inline their calls. Also, similar to before I think the compiler may be able to figure out how to construct the TileLayout just given:
`std::make_shared<TileLayout>(numberOfCols, numberOfRows, widthsForId(video, id), heightsForId(video, id));`

File: tasm/tiles/src/LayoutDatabase.cc [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+
+    std::experimental::filesystem::path directory = TileFiles::directoryForTilesInFrames(entry->path(), id, firstFrame, lastFrame);
+
+    return TileFiles::tileFilename(directory, tileNumber);
+}
+
+void LayoutDatabase::addTileLayout(const std::string &video, unsigned int id, unsigned int firstFrame, unsigned int lastFrame, const std::shared_ptr<TileLayout> tileLayout) {
+    addTileBasicLayout(video, id, firstFrame, lastFrame, tileLayout->numberOfColumns(), tileLayout->numberOfRows());
+    addTileLayoutWidths(video, id, tileLayout->widthsOfColumns());
+    addTileLayoutHeights(video, id, tileLayout->heightsOfRows());
```
Comment (maureendaum) [inline tasm/tiles/src/LayoutDatabase.cc:268]: I think we want to do this inside of a single transaction so that if any step fails, we won't end up with partial state. An example of this is in `SemanticIndexSQLiteBase::addBulkMetadata()`.

File: tasm/tiles/src/LayoutDatabase.cc [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+    int position = 0;
+    for (const auto& width : widths) {
+        ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 3, width));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 4, position));
+        ASSERT_SQLITE_DONE(sqlite3_step(select));
+        ASSERT_SQLITE_OK(sqlite3_reset(select));
+        ++position;
+    }
```
Comment (maureendaum) [inline tasm/tiles/src/LayoutDatabase.cc]: I think we also want this batch-insert to be inside of a transaction. I can think of a few options:
- We could add a comment saying that this function should only only be called within a transaction (e.g., from `addTileLayout()`)
- We could use sqlite's save points to create nested transactions: https://www.sqlite.org/lang_savepoint.html
- We could insert all of the values at once using an INSERT INTO … VALUES query (there are some examples here: https://www.sqlitetutorial.net/sqlite-insert/)

(Ditto for `addTileLayoutHeights`)

File: tasm/tiles/src/LayoutDatabase.cc [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+}
+
+unsigned int LayoutDatabase::totalHeight(const std::string &video) const {
+    // select all ids for this video from layouts
+    std::string query = "SELECT id FROM layouts WHERE video = ?";
+    sqlite3_stmt *select;
```
Comment (maureendaum) [inline tasm/tiles/src/LayoutDatabase.cc]: I think it's nice to check that the height is the same for all IDs, but I think it's something we can assume is true.  (Also, it will be super slow for long videos with lots of different layouts -- the complexity of this query will depend on the length of the video and number of layouts, rather than being constant). Then, a faster query would be "SELECT id FROM layouts WHERE video=? LIMIT 1".

File: tasm/tiles/src/LayoutDatabase.cc [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+
+    ASSERT_SQLITE_DONE(result);
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+
+    // for all ids, check that the total height is the same
+    assert(ids.size() > 0);
+
+    std::vector<unsigned int> heights = heightsForId(video, ids.at(0));
+    unsigned int totalHeight = std::accumulate(heights.begin(), heights.end(), 0);
+    for (unsigned int id : ids) {
```
Comment (maureendaum) [inline tasm/tiles/src/LayoutDatabase.cc]: If we don't verify that the height is the same for all IDs, then we could do this aggregation in the query: "SELECT sum(height) FROM heights WHERE video=? and id=?". On the other hand, it's nice to re-use the `heightsForId` function, despite it adding a bit to the complexity. It's up to you how you want to do this.

File: tasm/tiles/src/TiledVideoManager.cc [hunk 12]
```diff
@@ -6,77 +6,24 @@
...
+    return layouts_->tileLayoutForId(entry_->name(), id);
 }
 
 std::experimental::filesystem::path TiledVideoManager::locationOfTileForId(unsigned int tileNumber, int id) const {
-    std::scoped_lock lock(mutex_);
-    return TileFiles::tileFilename(directoryIdToTileDirectory_.at(id), tileNumber);
+    return layouts_->locationOfTileForId(entry_, tileNumber, id);
 }
 
 } // namespace tasm
```
Comment (maureendaum) [inline tasm/tiles/src/TiledVideoManager.cc:29]: It's great to see all of the deletions in this file!

File: tasm-test/src/TasmTest.cc [hunk 13]
```diff
@@ -2,6 +2,7 @@
 #include "Video.h"
 #include <gtest/gtest.h>
 #include "sqlite3.h"
+#include "LayoutDatabase.h"
```
Comment (maureendaum) [inline tasm-test/src/TasmTest.cc:5]: Is this import actually needed?

File: tasm-test/src/DatabaseTest.cc [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        std::string name = "name";
+
+        int tileNumber = 4;
+        unsigned int id = 2;
+        unsigned int firstFrame = 40;
+        unsigned int lastFrame = 59;
+        std::vector<unsigned int> heights{1, 2};
+        std::vector<unsigned int> widths{1, 2};
+        unsigned int numColumns = 2;
+        unsigned int numRows = 2;
```
Comment (maureendaum) [inline tasm-test/src/DatabaseTest.cc]: You could switch these to `widths.size(), heights.size()` like in the other tests.

File: tasm-test/src/DatabaseTest.cc [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+        std::shared_ptr<TileLayout> tileLayout = std::make_shared<TileLayout>(numColumns, numRows, widths, heights);
+
+        layoutDatabase->addTileLayout(name, id, firstFrame, lastFrame, tileLayout);
+
+        auto entry = std::make_shared<TiledEntry>(name);
+
+        std::experimental::filesystem::path expectedLocation = "";
+
+        std::experimental::filesystem::path tileLocation = layoutDatabase->locationOfTileForId(entry, tileNumber, id);
```
Comment (maureendaum) [inline tasm-test/src/DatabaseTest.cc]: What is this test actually checking? It doesn't look like expectedLocation and tileLocation are ever used.

**Review by maureendaum — COMMENTED** [review]
I think this looks great! Feel free to use/ignore any of the style/naming comments, and then let me know if you have any questions about the transaction or prepared statement stuff.

I think we can also probably delete the `TileConfiguration.proto` file at this point.

File: tasm-test/src/VideoManagerTest.cc [hunk 16]
```diff
@@ -10,8 +11,51 @@ using namespace tasm;
...
+        std::experimental::filesystem::copy("test-inputs/video-manager-test-resources/test-resources", config.catalogPath(), std::experimental::filesystem::copy_options::recursive);
+
+        LayoutDatabase::instance()->open();
+    }
+    void TearDown() {
+        // remove test resources
+        auto config = EnvironmentConfiguration::instance();
+        std::experimental::filesystem::remove_all(config.defaultLayoutDatabasePath());
+        std::experimental::filesystem::remove_all(config.catalogPath());
+        std::experimental::filesystem::remove_all(config.defaultLabelsDatabasePath());
```
Comment (maureendaum) [inline tasm-test/src/VideoManagerTest.cc:39]: A possible future enhancement could be to move these copy/remove calls that are common among the test cases to utility functions that can be called by each test.

File: tasm-test/src/TasmTest.cc [hunk 17]
```diff
@@ -90,7 +137,7 @@ TEST_F(TasmTestFixture, testSelectBird) {
 
 TEST_F(TasmTestFixture, testScanBirdsFullFrame) {
     tasm::TASM tasm(SemanticIndex::IndexType::XY, "/home/maureen/home_videos/birds_tasm.db");
-    auto selection = tasm.selectFrames("birds-birds", "bird", 0, 5, "birds");
+    auto selection = tasm.selectFrames("birdsincage-2x2", "bird", 0, 5, "birds");
```
Comment (maureendaum) [inline tasm-test/src/TasmTest.cc:135]: I forgot about copyright issues, and I don't believe any of the non-red10 videos have a copyright that allows us to embed it in this repository. Would it be possible to switch all uses of birds/elfuente/any other to use "red10"? Since we don't care about the actual contents of the videos, hopefully all of the tests still work. Let me know if you want me to generate red10 videos with the same dimensions/frame rate/number of frames as these other videos!

Comment (maureendaum) [reply tasm-test/src/TasmTest.cc:135]: If it's easiest for now, you can leave the tests as-is and just remove the videos from the commits.

**Review by maureendaum — APPROVED** [review]
Looks great! The only thing is the issue with the videos, but for now I recommend just removing them from the commit, and then the next time we meet we can figure out whether a generic video like red10 would be a suitable replacement.

File: tasm-test/src/LayoutBenchmarkTest.cc [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+    std::chrono::system_clock::time_point totalEnd = std::chrono::system_clock::now();
+
+    auto totalDuration = std::chrono::duration_cast<std::chrono::microseconds>(totalEnd - totalStart).count();
+    auto constructorDuration = std::chrono::duration_cast<std::chrono::microseconds>(constructorEnd - constructorStart).count();
+    auto functionDuration = std::chrono::duration_cast<std::chrono::microseconds>(functionEnd - functionStart).count();
+
+    // output data to csv file
+    printStat(outputFile, experiment, "total", videoName, frameNumber, iteration, totalDuration);
+    printStat(outputFile, experiment, "constructor", videoName, frameNumber, iteration, constructorDuration);
+    printStat(outputFile, experiment, videoName, "processing", frameNumber, iteration, functionDuration);
```
Comment (maureendaum) [inline tasm-test/src/LayoutBenchmarkTest.cc]: Should videoName and "processing" be switched?

File: tasm-test/src/LayoutBenchmarkTest.cc [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+    };
+    auto config = EnvironmentConfiguration::instance(EnvironmentConfiguration(options));
+
+    // initialize video info
+    std::string labelPaths[NUM_VIDEOS] LABEL_PATHS;
+    std::string videoPaths[NUM_VIDEOS] VIDEO_PATHS;
+    std::string labels[NUM_VIDEOS] VIDEO_LABELS;
+    std::string names[NUM_VIDEOS] VIDEO_NAMES;
+    int numFrames[NUM_VIDEOS] NUM_FRAMES;
+    int ids[NUM_VIDEOS] IDS;
```
Comment (maureendaum) [inline tasm-test/src/LayoutBenchmarkTest.cc]: Alternatively, you could define all of these constants as their actual types at the top of the file and just use them here.

File: tasm-test/src/LayoutBenchmarkTest.cc [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+#include <functional>
+
+using namespace tasm;
+
+#define VIDEO_PATHS {"/home/maureen/red_videos/red_6sec_2k.mp4", "/home/maureen/red_videos/red_900sec_2k.mp4"}
+#define LABEL_PATHS {"/home/maureen/NFLX_dataset/detections/labels/birdsincage.db", "/home/maureen/visualroad_tiling/labels_yolo/traffic-2k-001_yolo.db"}
+#define VIDEO_LABELS {"bird", "car"}
+#define VIDEO_NAMES {"birdsincage", "traffic-2k-001_yolo"}
+#define NUM_FRAMES {180, 27000}
+#define IDS {197, 8476}
```
Comment (maureendaum) [inline tasm-test/src/LayoutBenchmarkTest.cc]: It would be helpful to add a comment specifying how we picked these IDs if it's important so future us knows how to update them if we modify the tests.

File: tasm-test/src/LayoutBenchmarkTest.cc [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+    for (int i = 0; i < NUM_VIDEOS; i++) {
+        std::string labelPath = labelPaths[i];
+        std::string videoPath = videoPaths[i];
+        std::string label = labels[i];
+        std::string name = names[i];
+        int frames = numFrames[i];
+        int id = ids[i];
+
+        std::experimental::filesystem::remove(config.defaultLayoutDatabasePath());
+        LayoutDatabase::instance()->open();
```
Comment (maureendaum) [inline tasm-test/src/LayoutBenchmarkTest.cc:170]: Are these the only two lines that are specific to the layout database instance?

File: tasm-test/src/LayoutBenchmarkTest.cc [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+#define NUM_ITERATIONS 3
+#define FUNCTION_REPETITIONS 50 // for functions that are too fast to measure without repetition
+#define SEP ","
+#define IMPLEMENTATION "layout database"
+#define OUTFILE_NAME "layout_database_benchmark.csv"
+
+void printStat(std::ofstream &file, std::string experiment, std::string step, std::string video, int frame,
+               int iteration, long time) {
+    file << IMPLEMENTATION << SEP << experiment << SEP << step << SEP << video << SEP << frame << SEP << iteration << SEP << time << "\n";
+}
```
Comment (maureendaum) [inline tasm-test/src/LayoutBenchmarkTest.cc:32]: We may want to also flush the ofstream so that even if something fails and the tests don't finish, we should hopefully have partial results in the output file.

File: tasm-test/src/LayoutBenchmarkTest.cc [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+
+#define VIDEO_PATHS {"/home/maureen/red_videos/red_6sec_2k.mp4", "/home/maureen/red_videos/red_900sec_2k.mp4"}
+#define LABEL_PATHS {"/home/maureen/NFLX_dataset/detections/labels/birdsincage.db", "/home/maureen/visualroad_tiling/labels_yolo/traffic-2k-001_yolo.db"}
+#define VIDEO_LABELS {"bird", "car"}
+#define VIDEO_NAMES {"birdsincage", "traffic-2k-001_yolo"}
+#define NUM_FRAMES {180, 27000}
+#define IDS {197, 8476}
+#define NUM_VIDEOS 2
+#define NUM_ITERATIONS 3
+#define FUNCTION_REPETITIONS 50 // for functions that are too fast to measure without repetition
```
Comment (maureendaum) [inline tasm-test/src/LayoutBenchmarkTest.cc]: Do you think we should also include FUNCTION_REPETITIONS in the output file? (With a value of 1 for the non-repeated ones) My thought is that it's possible this could change over time, so it could be useful to know for a specific set of results what its value was.

**Comment by kirstenmg** [comment]
Benchmarking showed that this implementation makes the constructor of TiledVideoManager much faster and less impacted by video size than the metadata file version. However, tileLayoutForId, locationOfTileForId, and tileLayoutIdsForFrame are slower with this version. A high (~5,000 to 10,000) number of these queries would outweigh the benefit of the faster constructor. Video storage performance is similar for both implementations.
